tools/backups2datalad/annex.py

```diff
@@ -31,10 +32,34 @@ class AsyncAnnex(anyio.abc.AsyncResource):
         init=False, default_factory=lambda: defaultdict(anyio.Lock)
     )
 
-    async def aclose(self) -> None:
-        for p in [self.pfromkey, self.pexaminekey, self.pwhereis, self.pregisterurl]:
-            if p is not None:
-                await p.aclose()
+    async def __aenter__(self) -> AsyncAnnex:
+        return self
+
+    async def __aexit__(
+        self,
+        exc_type: Optional[type[BaseException]],
+        _exc_val: Optional[BaseException],
+        _exc_tb: Optional[TracebackType],
+    ) -> None:
+        if exc_type is None:
+            for p in [
+                self.pfromkey,
+                self.pexaminekey,
+                self.pwhereis,
+                self.pregisterurl,
+            ]:
+                if p is not None:
+                    await p.aclose()
```

Comment:
neither here nor below we `self.pWHATEVER = None` after closing. Is it guaranteed that we aren't reusing those `self.pWHATEVER`s somehow and thus ending up trying to access closed things?

should may be `force_close` do the `.aclose` first and then if not dead after a bit proceed with terminate etc, otherwise may be we would be getting some run aways here?

Reply:
* This code is only called on exit from an `AsyncAnnex`'s `async with` block.  Neither instance of `AsyncAnnex` in the code is used in more than one block, so there's no reuse.
* Done: 7ba01c3

Reply:
- ok, let's hope there is no reuse
- why don't we use now `.force_close()` here too?  but then do we really need `if exc_type is None` above or we could just use the `else:` block?

Reply:
I am still interested in the answer here but will merge for now to make use of this PR

Reply:
If `exc_type is None` is true, then there was no error and everything completed normally, so we should expect git-annex to terminate right after closing stdin.